core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala

```diff
@@ -283,6 +285,18 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
     val stage = stageSubmitted.stageInfo
     activeStages(stage.stageId) = stage
     pendingStages.remove(stage.stageId)
+
+    // If a stage retries again, it should be removed from completedStageIndices set
+    for (
+      activeJobsDependentOnStage <- stageIdToActiveJobIds.get(stage.stageId);
+      jobId <- activeJobsDependentOnStage;
+      jobData <- jobIdToData.get(jobId)
+    ) {
+      if (jobData.completedStageIndices.contains(stage.stageId)) {
```

Comment:
Oh I see you made this change already. I don't think you have to check `contains` to `remove` it; it won't do anything if not present. Why the loop here -- can this not happen in the same loop farther down?